docs/lofi-hifi-backends.md

```diff
@@ -0,0 +1,279 @@
+
+`ilab model train phased` will opt to use torchrun under the hood, stepping away from the transformers library a bit. Though, research in future versions will most likely lead to a convergance between the two since transformers paired with torch can do most of the tasks we manually shell out to torchrun and torch.distriuted for. 
+
+The DeepSpeed config currently used for `ilab model train phased` will have more hardcoded values, until flags are added in later versions allowing users to configure things like the trianing batch size, offloading optimizer etc. with PEFT most of that is handeled using the `auto` keyword that transformers can funnel defaults into.
+
+However, if `--device=cpu` the offload optimizer and other parts of the deepspeed config will not be relevant.
+
+### ilab data generate hifi
+
+This command would take something like the following arguments
+
+--num-samples=int
+--num-grounded-questions=int
+--num-gen-proc=int
+--num-util-proc=int (or is this for mixing?)
+
+
+### ilab data generate lofi
+
+This command would be the same as the existing `ilab generate`
+
+### ilab data mix
+
+This command would take something like the following arguments
+
+--num-util-proc=int
+--output-dir=str (defaults to generated/mixed)
+--knowledge-recipes=[]str (path to yaml)
+--skill-recipes=[]str (path to yaml)
+
+* Do we need an `ilab recipe` cmd? *
+
+
+## Workflows
+
+### ilab model train peft
+
+A user on a Desktop with a consumer GPU (assume RTX 20/30 series) would run something like
+
+`ilab model train peft --device=cuda --quantize --accelerator=deepspeed --num-epochs=5`
+
+This would 
+1. load the model in 4-bit-quantized form onto the GPUs vram
+2. setup the transformers trainer with this model, and with a hardcoded deepspeed config ilab would come with
+3. train for 5 epochs using AdamW as the optim and deepspeed on top of that.
+4. give you a model in safetensors format (we cannot convert a quantized safetensor model)
+
+
+The big advantage here is faster and higher fidelity training than currently exists in the CLI because of deepspeed (or fsdp). The user could even set this up for multi GPU or multi system support with future ilab enhancements.
+
+
+`ilab model train peft --device=cuda --accelerator=deepspeed --num-epochs=5`
+
+This would 
+1. load the model in its full form to the GPUs vram (multiple GPUs in this case)
+2. setup the transformers trainer with this model, and with a hardcoded deepspeed config ilab would come with
+3. train for 5 epochs using AdamW as the optim and deepspeed on top of that.
+4. give you a model in safetensors format or GGUF format since this is not a bitsnbytes model.
+
+
+The big advantage here is faster and higher fidelity training than currently exists in the CLI because of deepspeed (or fsdp). The user could even set this up for multi GPU or multi system support with future ilab enhancements.
```

Comment:
The benefits from this approach look similar to the ones when using `--quantize`, can we add more on the gains?